### PR TITLE
[MRG] OpenBCI

### PR DESCRIPTION
**Pull request by teonlamont**
This is wip for reading openbci data to mne-python. Tests are needed. 

This handles missing data packets which can happen when wirelessly transmitting data from the EEG. Currently a simple average is done between the data packets that are missing. Possibly use something a bit more sophisticated: interpolation or resample for missing data?

### CODE REVIEW

**Comment by agramfort** [comment]
Now I am curious why you look into this :)

**Comment by teonlamont** [comment]
:) I recently joined [NeuroTechX](http://neurotechx.com), NYC chapter. [OpenBCI](http://openbci.com) is based in Brooklyn and I've been working with @conorrussomanno, its co-founder and also the NeuroTechNYC organizer, on integrating their tech with mne-python, making the ultimate open-source amalgamation ;)

**Comment by conorrussomanno** [comment]
Can't wait for mne-python to be pulling data from OpenBCI! :D

Conor

-- 
Co-Founder & CEO | OpenBCI <http://www.openbci.com/>
Part-Time Faculty | Parsons MFADT
<http://www.newschool.edu/parsons/mfa-design-technology/>
Twitter: @russomanno15 <https://twitter.com/russomanno15>

**Comment by agramfort** [comment]
hehe great ! I am very much looking forward to it too.

**Comment by teonlamont** [comment]
@agramfort or @Eric89GXL, what tests should I add for this reader? in the past, my tests compared the reading of the data to other package (e.g. eeglab) but I think we're the only ones at the moment that are supporting this format. I just added a copy of the test data.

File: mne/io/bci/bci.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+    reference : None | str
+        Name of the electrode which served as the reference in the recording.
+        If a name is provided, a corresponding channel is added and its data
+        is set to 0. This is useful for later re-referencing. The name should
+        correspond to a name in elp_names. Data must be preloaded.
+    scale : float
+        The scaling factor for EEG data. Units for MNE are in volts.
+        BCI data are typically stored in microvolts. Default scale
+        factor is 1e-6.
+    sfreq : int
```
Comment (Eric89GXL) [inline mne/io/bci/bci.py]: float

File: mne/io/bci/bci.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    -----
+    .. versionadded:: 0.10.0
+
+    See Also
+    --------
+    mne.io.Raw : Documentation of attribute and methods.
+    """
+    @verbose
+    def __init__(self, input_fname, montage=None, eog=None,
+                 misc=(-3, -2, -1), reference=None, scale=1e-6, sfreq=250,
```
Comment (Eric89GXL) [inline mne/io/bci/bci.py]: 250.

File: mne/io/bci/bci.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+                 preload=True, verbose=None):
+
+        if not eog:
+            eog = list()
```
Comment (Eric89GXL) [inline mne/io/bci/bci.py]: FYI a one-line equivalent is `eog = eog or list()` but it's probably too obfuscated for use here

**Comment by Eric89GXL** [comment]
Are there any readers e.g. for MATLAB? It would be good to read with one of those, and write out to a .mat file or equivalent. Then load yours and compare the result. Then save yours to disk as FIF, load it, and compare again. Then check coverage and see if you can get 100%, including the interpolation stuff (get new file and repeat above if it is not missing samples). That would be a good start.

**Comment by teonlamont** [comment]
there's only one thing I haven't been able to reckon with: we've fixed the way to fix skipped data packets but we don't have a good way to update the number of total samples, or  `last_samp`. you can only do it after you've read in all the data. so when `preload=False`, `nsamps` may be incorrect.

**Comment by teonlamont** [comment]
now it's ready

**Comment by agramfort** [comment]
you'll need to rebase.

also I'd use OpenBCI for class and open_bci in functions/filenames. BCI is too generic

**Comment by agramfort** [comment]
data are stored in plan text?

**Comment by teonlamont** [comment]
yeah, as of now. we're also looking other format for future versions (edf/bdf/gdf/xdf)

**Comment by agramfort** [comment]
oh boy... it implies huge files and limited numerical precision. You need
to fix this.

**Comment by teonlamont** [comment]
totally agree. working on it now :)

**Comment by Eric89GXL** [comment]
It would probably not be terribly difficult to write it to HDF5 if one were
so inclined. But maybe we have enough formats already...

Is there a format like nifti/gifti that we should be using for MEG data but
aren't?

**Comment by agramfort** [comment]
to me the OpenBCI folks should export to EDF rather than inventing a new
format.

that's the simple approach. Everybody can read edf.

**Comment by teonlamont** [comment]
yeah, that was my suggestion as well. we're having the discussion on neurotechx [slack] (https://neurotechx.slack.com/messages/eegdata/) right now. 

at the moment, the format has been this for the past year. 

I will be working with them tomorrow and I'll see if we can get their acquisition software to export an EDF format.

**Comment by agramfort** [comment]
the benefit is that you would not have to maintain another IO piece of code
:)

**Comment by teonlamont** [comment]
> Is there a format like nifti/gifti that we should be using for MEG data but
> aren't?

I know that some people from EEGLab have started working on [XDF](https://github.com/sccn/xdf), which is supposed to be another format to deal with these data.

The limitation of EDF is 16-bit precision, but @agramfort already mentioned it's supported by the number of packages.

One possibility for this format is that it could be hosted on OpenBCI git account to be backwards compatible (the dirtiest word in programming) with the current data format so that people can enjoy our package and work toward a simpler/standard format.

**Comment by teonlamont** [comment]
> the benefit is that you would not have to maintain another IO piece of code
> :)

@agramfort don't you like having yet another standard format ;)

**Comment by agramfort** [comment]
I am getting old and I've stopped being enthusiast about new file formats :)

**Comment by teonlamont** [comment]
closing this and migrating this function to OpenBCI. once we get the data in a standard format, I will update the closed PR, and write a blurb about MNE-compatibility with the system

**Comment by teonlamont** [comment]
can follow at https://github.com/OpenBCI/OpenBCI_Processing/pull/28

**Pull request closed without merging**